internal/controller/namespacelabel_controller.go

```diff
@@ -0,0 +1,242 @@
+	var namespaceLabel labelsv1.Namespacelabel
+	if err := r.Get(ctx, req.NamespacedName, &namespaceLabel); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("failed to get namespace label: %w", err))
+	}
+
+	if !namespaceLabel.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.handleDeletion(ctx, &namespaceLabel); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to handle deletion: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ensureFinalizer(ctx, &namespaceLabel); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	protectedLabels, err := labels.LoadProtectedLabels(r.Log)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to load the protected labels list: %w", err)
+	}
+
+	namespace, err := r.fetchNamespace(ctx, namespaceLabel.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	updatedLabels, skippedLabels, duplicateLabels := r.processLabels(namespace, &namespaceLabel, protectedLabels)
+
+	for key, value := range updatedLabels {
+		namespace.Labels[key] = value
+	}
+
+	if err := r.updateNamespace(ctx, namespace); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.updateStatus(ctx, &namespaceLabel, updatedLabels, skippedLabels, duplicateLabels); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update Namespacelabel status: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NamespacelabelReconciler) SetCondition(namespaceLabel *labelsv1.Namespacelabel, conditionType string, status metav1.ConditionStatus, reason, message string) {
+	r.Log.Info("Setting condition", "type", conditionType, "status", status, "reason", reason)
+
+	condition := metav1.Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastTransitionTime: metav1.Now(),
+	}
+
+	meta.SetStatusCondition(&namespaceLabel.Status.Conditions, condition)
+}
+
+func (r *NamespacelabelReconciler) handleDeletion(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel) error {
+	r.Log.Info("Handling deletion for Namespacelabel", "namespace", namespaceLabel.Namespace)
+	if err := finalizer.Cleanup(ctx, r.Client, namespaceLabel, r.Log); err != nil {
+		return fmt.Errorf("failed to clean up labels during finalizer: %w", err)
+	}
+	return nil
+}
+
+func (r *NamespacelabelReconciler) fetchNamespace(ctx context.Context, namespaceName string) (*corev1.Namespace, error) {
+	var namespace corev1.Namespace
+	if err := r.Get(ctx, types.NamespacedName{Name: namespaceName}, &namespace); err != nil {
+		r.Log.Error(err, "Failed to get namespace")
+		return nil, fmt.Errorf("failed to get namespace: %w", err)
+	}
+	return &namespace, nil
+}
+
+func (r *NamespacelabelReconciler) updateNamespace(ctx context.Context, namespace *corev1.Namespace) error {
+	if err := r.Update(ctx, namespace); err != nil {
+		return fmt.Errorf("failed to update namespace: %w", err)
+	} else {
+		return nil
+	}
```

Comment:
Redundant `else`. This entire function can just be in the caller instead of in its own function anyway.

If you want to keep it as a separate function then you can generalize it and use the `client.Object` interface instead of passing a `namespace *corev1.Namespace` as a parameter to the function